#! /usr/bin/env python

# Declare parameters that control stereo processing

PACKAGE='stereo_image_proc'

from dynamic_reconfigure.parameter_generator_catkin import *

gen = ParameterGenerator()

# disparity block matching pre-filtering parameters
gen.add("prefilter_cap",  int_t, 0, "Bound on normalized pixel values", 31, 1, 63)

# disparity block matching correlation parameters
gen.add("correlation_window_size", int_t, 0, "SAD correlation window width, pixels", 15, 5, 255)
gen.add("min_disparity",           int_t, 0, "Disparity to begin search at, pixels (may be negative)", 0, -128, 128)
gen.add("disparity_range",         int_t, 0, "Number of disparities to search, pixels", 64, 32, 128)
# TODO What about trySmallerWindows?

# disparity block matching post-filtering parameters
# NOTE: Making uniqueness_ratio int_t instead of double_t to work around dynamic_reconfigure gui issue
gen.add("uniqueness_ratio",  double_t, 0, "Filter out if best match does not sufficiently exceed the next-best match", 15, 0, 100)
gen.add("texture_threshold", int_t,    0, "Filter out if SAD window response does not exceed texture threshold", 10, 0, 10000)
gen.add("speckle_size",      int_t,    0, "Reject regions smaller than this size, pixels", 100, 0, 1000)
gen.add("speckle_range",     int_t,    0, "Max allowed difference between detected disparities", 4, 0, 31)

# disparity semi-global specific block matching parameters
gen.add("p1",  		int_t,  0, "..", 600, 0, 10000)
gen.add("p2", 		int_t,  0, "..", 2400, 0, 10000)
gen.add("max_diff", int_t,  0, "..", 10, 0, 1000)
gen.add("full_dp",  bool_t, 0, "..", False)

# First string value is node name, used only for generating documentation
# Second string value ("Disparity") is name of class and generated
#    .h file, with "Config" added, so class SGDisparityConfig
exit(gen.generate(PACKAGE, "stereo_image_proc", "SGDisparity"))
